pkg/util/tracing/tracer.go

```diff
@@ -156,179 +157,65 @@ func (t *Tracer) getShadowTracer() *shadowTracer {
 	return (*shadowTracer)(atomic.LoadPointer(&t.shadowTracer))
 }
 
-type recordableOption struct{}
-
-// Apply is part of the opentracing.StartSpanOption interface.
-func (recordableOption) Apply(*opentracing.StartSpanOptions) {}
-
-// Recordable is a StartSpanOption that forces creation of a real Span.
-//
-// When tracing is disabled all spans are noopSpans; these spans aren't
-// capable of recording, so this option should be passed to StartSpan if the
-// caller wants to be able to call StartRecording on the resulting Span.
-var Recordable opentracing.StartSpanOption = recordableOption{}
-
-// StartSpan is part of the opentracing.Tracer interface.
-//
-// Avoid using this method in favor of Tracer.StartRootSpan() or
-// tracing.StartChildSpan() (or higher-level methods like EnsureContext() or
-// AmbientContext.AnnotateCtxWithSpan()) when possible. Those are more efficient
-// because they don't have to use the StartSpanOption interface and so a bunch
-// of allocations are avoided. However, we need to implement this standard
-// StartSpan() too because grpc integrates with a generic opentracing.Tracer.
-func (t *Tracer) StartSpan(operationName string, opts ...opentracing.StartSpanOption) *Span {
+// StartSpan starts a Span.
+func (t *Tracer) StartSpan(operationName string, os ...SpanOption) *Span {
 	// Fast paths to avoid the allocation of StartSpanOptions below when tracing
 	// is disabled: if we have no options or a single SpanReference (the common
-	// case) with a noop context, return a noop Span now.
-	if len(opts) == 1 {
-		if o, ok := opts[0].(opentracing.SpanReference); ok {
-			if o.ReferencedContext.(*SpanContext).IsNoop() {
+	// case) with a black hole span, return a noop Span now.
+	if len(os) == 1 {
+		switch o := os[0].(type) {
+		case *parentOption:
+			if (*Span)(o).IsBlackHole() {
+				return t.noopSpan
+			}
+		case *remoteParentOption:
+			if (*SpanMeta)(o).isNilOrNoop() {
 				return t.noopSpan
 			}
 		}
 	}
-
-	if len(opts) == 0 && !t.AlwaysTrace() {
+	if len(os) == 0 && !t.AlwaysTrace() {
 		return t.noopSpan
 	}
 
-	var sso opentracing.StartSpanOptions
-	var recordable RecordableOpt
-	var logTags *logtags.Buffer
-	for _, o := range opts {
-		switch to := o.(type) {
-		case recordableOption:
-			recordable = true
-		case *logTagsOption:
-			// NOTE: calling logTagsOption.Apply() would be functionally equivalent,
-			// but less efficient.
-			logTags = (*logtags.Buffer)(to)
-		default:
-			o.Apply(&sso)
-		}
+	var opts SpanOptions
+	// TODO(during review): `./scripts/goescape ./pkg/util/tracing/tracer.go` claims that
```

Comment:
@nvanbenschoten here's a doozy for you.

Reply:
We never got to writing that escape linter, right? I had a first attempt at https://github.com/tbg/goescape, but I think we both agreed that a better approach would've been to have an inlineable `func(interface{})` that we can find the arguments to and lint that way.

Reply:
Yeah, this one's bizarre. I don't see how `opts` can not escape to the heap. You're likely right that this is because goescape is only compiling this single package, though I'm still surprised that the compiler is smart enough to determine that `opts` doesn't escape in any implementation of `SpanOption`. Have you tried making it escape in one of the implementations and seeing what happens?

Reply:
Heh, the answer is simple: I missed the line. It's right there:

```
$ ./scripts/goescape ./pkg/util/tracing/tracer.go | grep tracer.go:181
./tracer.go:181:6: moved to heap: opts
```

Sorry about the noise.